server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,308 @@
+    public SnapshotShardSizeInfo snapshotShardSizes() {
+        return new SnapshotShardSizeInfo(knownSnapshotShardSizes);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
+            final Set<SnapshotShard> onGoingSnapshotRecoveries = new HashSet<>();
+
+            int unknownShards = 0;
+            for (ShardRouting shardRouting : event.state().routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+                if (shardRouting.primary() && shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+                    final RecoverySource.SnapshotRecoverySource snapshotRecoverySource =
+                        (RecoverySource.SnapshotRecoverySource) shardRouting.recoverySource();
+                    final SnapshotShard snapshotShard = new SnapshotShard(snapshotRecoverySource.snapshot(),
+                        snapshotRecoverySource.index(), shardRouting.shardId());
+                    onGoingSnapshotRecoveries.add(snapshotShard);
+
+                    // check if already populated entry
+                    if (knownSnapshotShardSizes.containsKey(snapshotShard) == false) {
+                        // check if already fetching snapshot info in progress
+                        if (unknownSnapshotShards.add(snapshotShard)) {
+                            queue.add(snapshotShard);
+                            unknownShards += 1;
+                        }
+                    }
+                }
+            }
+
+            // Clean up keys from knownSnapshotShardSizes that are no longer needed for recoveries
+            synchronized (this) {
+                isMaster = true;
+                ImmutableOpenMap.Builder<SnapshotShard, Long> newSnapshotShardSizes = null;
+                for (ObjectCursor<SnapshotShard> shard : knownSnapshotShardSizes.keys()) {
+                    if (onGoingSnapshotRecoveries.contains(shard.value) == false) {
+                        if (newSnapshotShardSizes == null) {
+                            newSnapshotShardSizes = ImmutableOpenMap.builder(knownSnapshotShardSizes);
+                        }
+                        newSnapshotShardSizes.remove(shard.value);
+                    }
+                }
+                if (newSnapshotShardSizes != null) {
+                    knownSnapshotShardSizes = newSnapshotShardSizes.build();
+                }
+            }
+
+            final int nbFetchers = Math.min(unknownShards, semaphore.getMaxPermits());
+            for (int i = 0; i < nbFetchers; i++) {
+                fetchNextSnapshotShard();
+            }
+        } else {
+            synchronized (this) {
+                // information only needed on current master
+                knownSnapshotShardSizes = ImmutableOpenMap.of();
+                isMaster = false;
+            }
+        }
+    }
+
+    private void fetchNextSnapshotShard() {
+        if (semaphore.tryAcquire()) {
```

Comment:
Do we really need a `Semaphore` here? Can't we just have these tasks increment a counter with `incrementAndGet` whenever we try to start one and if it's below the `current limit + 1` we start a task? Then we only need to decrement the counter when we poll a `null` from the queue and it's all a lot easier isn't it?

Reply:
Something like https://github.com/elastic/elasticsearch/pull/61906/commits/6c59655bd4c6d5c6ccc19c1a9c9324a3c170e910?